Text/TextQualityWatchdog/Util/TikaExtract.py

```diff
@@ -0,0 +1,159 @@
+
+# Metrics directory where we save metrics used to summarize the run
+METRICS_DIR = 'Metrics'
+
+# Retrieve the connection string for use with the application. The storage
+# connection string is stored in an environment variable on the machine
+# running the application called AZURE_STORAGE_CONNECTION_STRING.
+connect_str = os.getenv('AZURE_STORAGE_CONNECTION_STRING')
+
+# Create the BlobServiceClient object which will be used to create a container client
+blob_service_client = BlobServiceClient.from_connection_string(connect_str)
+
+# Get the container client corresponding to the commoncrawl container
+cont_client = blob_service_client.get_container_client('commoncrawl3-refetched')
+
+# Get the list of all the blobs inside the commoncrawl container
+blob_list = cont_client.list_blobs()
+
+# Read the high water mark
+last_read_file_name = None
+last_read_file_path = APP_ROOT_STR + "\\" + METRICS_DIR + "\\lastread.txt"
+if os.path.exists(last_read_file_path):
+    with open(last_read_file_path, "r") as last_read_file:
+        last_read_file_name = last_read_file.readline()
+
+# Read the metrics dictionary, or create it if it does not exist
+metrics_dict = None
+metrics_dict_file_path = APP_ROOT_STR + "\\" + METRICS_DIR + "\\dataset_metrics.pkl"
+if os.path.exists(metrics_dict_file_path):
+    with open(metrics_dict_file_path, "rb") as metrics_dict_file:
+        metrics_dict = pickle.load(metrics_dict_file)
+else:
+    metrics_dict = {
+        "success" : 0, # English files that had text successfully extracted
+        "nonEng" : 0, # Files that were not in English
+        "failed" : 0, # Files that failed to have text extracted
+        "total" : 0 # Total count of the files
+    }
+
+# Process each blob
+for cur_blob in blob_list:
+    try:
+        # Get the name of the current blob
+        blobname = cur_blob.name
+        filename = blobname.split('/')[1]
+        print(blobname)
+
+        # Iterate until we hit the high water mark
+        if last_read_file_name != None and filename < last_read_file_name:
+            print("skip!")
+            continue
+
+        # Get the blob client for the current blob
+        blob_client = cont_client.get_blob_client(blobname)
+
+        # Get the path to save the unstructured file to
+        download_file_path = APP_ROOT_STR + "\\" + DOWNLOAD_DIR + "\\" + filename
+
+        # Get the full path to the destination file
+        extracted_file_path = APP_ROOT_STR + "\\" + DEST_DIR + "\\" + filename
+
+        # Get the full path to the metadata file
+        metadata_file_path = APP_ROOT_STR + "\\" + METADATA_DIR + "\\" + filename
+
+        # Download the blob to a local file
+        with open(download_file_path, "wb") as download_file:
+            download_file.write(blob_client.download_blob().readall())
+        print("Downloaded blob to \n\t" + download_file_path)
+
+        # Invoke tika server to perform extraction
+        parsedFile = parser.from_file(download_file_path)
```

Comment:
Given everything that can happen with Tika, can we add a little more around this to support timeouts, error handling, retries, etc.?

Reply:
That's a good call. From my limited experience with Tika, I've found that a lot of the time, simply rerunning the parser will resolve the errors that it had the first time. So, I added a PARSER_MAX_RETRIES constant to control how many times to retry the Tika parser if it fails to extract text, as well as more explicit exception handling logic.